### PR TITLE
Importing core and Running a slice as one process and Prime/Region/Zone threads

### PR DESCRIPTION
**Pull request by gameofpointers**

### CODE REVIEW

**Comment by wizeguyy** [comment]
@gameofpointers, your first commit ([Imports all flags from go-quai](https://github.com/dominant-strategies/go-quai-libp2p/pull/30/commits/39006c2e0883b4db77aaaf4e48a7383391ab4e14)) should really be two commits. Please split it into two:
1) change how flags are processed
2) add new flags...

File: cmd/utils/defaults.go [hunk 1]
```diff
@@ -0,0 +1,549 @@
...
+	},
+	TxPool:      TxPoolConfig{},
+	RPCGasCap:   50000000,
+	GPO:         FullNodeGPOConfig,
+	RPCTxFeeCap: 1, // 1 ether
+	DomUrl:      "ws://127.0.0.1:8546",
+	SubUrls:     []string{"ws://127.0.0.1:8546", "ws://127.0.0.1:8546", "ws://127.0.0.1:8546"},
+}
+
+type EthConfig struct {
```
Comment (wizeguyy) [inline cmd/utils/defaults.go:76]: This file confuses me. For example, why is the `EthConfig` type defined in `utils/defaults.go`?  `EthConfig` is not a util. It should not be defined in the utils module. Presumably it is a config for consensus, so it should be defined there.

Same argument can be made for all the other types defined in this file

File: cmd/utils/flags.go [hunk 2]
```diff
@@ -31,6 +31,120 @@ var NodeFlags = []Flag{
 	MaxPeersFlag,
 	LocationFlag,
 	SoloFlag,
+	DBEngineFlag,
+	KeystoreDirFlag,
+	NoUSBFlag,
+	USBFlag,
```
Comment (wizeguyy) [inline cmd/utils/flags.go]: Please don't add flags which aren't in use today. This PR should only add flags which are actively being used by the consensus module. Future modules can add flags as need arises.

File: cmd/utils/flags.go [hunk 3]
```diff
@@ -849,8 +846,48 @@ var (
 		Value: EthConfigDefaults.DomUrl,
 		Usage: "Subordinate chain websocket urls" + generateEnvDoc("sub.urls"),
 	}
+	CoinbaseAddressFlag = Flag{
+		Name:  "coinbase",
+		Value: "./coinbases.json",
```
Comment (wizeguyy) [inline cmd/utils/flags.go:625]: This default value is incorrect. It should be a relative path that is modified by the configDir flag.

in-fact, coinbase addresses should just be a section of the config.toml. There should not be a separate coinbases file for those.

File: coinbases.json
Comment (wizeguyy) [inline coinbases.json:1]: this file should not exist

File: Makefile [hunk 4]
```diff
@@ -32,4 +32,9 @@ protogen:
 ## build the go-quai binary
 build:
 	@echo "Building go-quai"
-	@go build -o go-quai ./cmd/go-quai/main.go
\ No newline at end of file
+	@go build -o build/bin/go-quai main.go
+
+run:
```
Comment (wizeguyy) [inline Makefile:41]: please do not add a run command to the makefile. The makefile is for building only.

In the future, if we need more complex run logic than just `go run ...`, we can consider writing a run script, but that is not needed today

File: README.md [hunk 5]
```diff
@@ -14,13 +14,13 @@ Once you have the necessary [prerequisites](#prerequisites), clone the `go-quai`
...
 cd go-quai
-go build -o ./build/go-quai main.go
+make go-quai
 ```
 
 ### Running a node
 To run a node with default options, simply execute:
 ```shell
-./build/go-quai start
+make run
```
Comment (wizeguyy) [inline README.md:23]: no

File: build/ci.go [hunk 6]
```diff
@@ -0,0 +1,263 @@
...
+	case "lint":
+		doLint(os.Args[2:])
+	default:
+		log.Fatal("unknown command ", os.Args[1])
+	}
+}
+
+// Compiling
+
+func doInstall(cmdline []string) {
```
Comment (wizeguyy) [inline build/ci.go:124]: We do not need/want this. There is no such thing as an "install" for us.

File: internal/build/archive.go
Comment (wizeguyy) [inline internal/build/archive.go:1]: Why is this file needed?

File: internal/build/download.go
Comment (wizeguyy) [inline internal/build/download.go:1]: why is this file needed?

File: internal/build/env.go
Comment (wizeguyy) [inline internal/build/env.go:1]: why is this file needed?

File: internal/build/gotool.go
Comment (wizeguyy) [inline internal/build/gotool.go:1]: Why is this file needed?

File: internal/build/pgp.go
Comment (wizeguyy) [inline internal/build/pgp.go:1]: Why is this file needed?

File: internal/build/util.go
Comment (wizeguyy) [inline internal/build/util.go:1]: Why is this file needed?

File: common/timedcache/timedcache.go
Comment (wizeguyy) [inline common/timedcache/timedcache.go:1]: I don't think this module is used anymore, right? no sense bringing it in

File: core/.gitignore
Comment (wizeguyy) [inline core/.gitignore:1]: I don't think you meant to leave a .gitignore here, right?

File: network.env
Comment (wizeguyy) [inline network.env:1]: network.env should not exist in this repo

File: params/bootnodes.go
Comment (wizeguyy) [inline params/bootnodes.go:1]: This conflicts with the bootnodes we've already defined. Please do not add this file, and instead use the bootnode definition options already defined by the P2P work

File: cmd/utils/cmd.go [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+	"github.com/dominant-strategies/go-quai/params"
+	"github.com/dominant-strategies/go-quai/quaistats"
+	"github.com/spf13/viper"
+	"io"
+	"os"
+	"runtime"
+	"time"
+)
+
+type quaistatsConfig struct {
```
Comment (wizeguyy) [inline cmd/utils/cmd.go:22]: `quaistatsConfig` and `quaiConfig` should not be defined in the utils module

File: cmd/utils/defaults.go (hunk 1 above)
Comment (gameofpointers) [reply cmd/utils/defaults.go:76]: I agree, it shouldn't be in this fine @robschleusner just put it in this file to get it compiled since we didnt have the structure of backend in, I will move it to a proper folder

File: cmd/utils/flags.go (hunk 2 above)
Comment (gameofpointers) [reply cmd/utils/flags.go]: Please look at the last few commits, this was just adding and then i removed the flags that were not in use. I guess will just squash the commits then

File: cmd/utils/flags.go (hunk 3 above)
Comment (gameofpointers) [reply cmd/utils/flags.go:625]: @robschleusner Can you see how to make this change?

File: coinbases.json
Comment (gameofpointers) [reply coinbases.json:1]: Yes, will be removed and done what you suggested @robschleusner

File: Makefile (hunk 4 above)
Comment (gameofpointers) [reply Makefile:41]: what do you mean not needed, there are bunch of flags we are using to start, I can remove it and just update the makefile with the commands under the run

Comment (gameofpointers) [reply Makefile:41]: And also we cannot do `go run`. what is the point of building if we are running the main.go, we need to use the build and not do `go run`

File: README.md (hunk 5 above)
Comment (gameofpointers) [reply README.md:23]: yeah, will just pass in the flags, with the build. I just made the make run command for the ease of testing. don't have an opinion on this matter. will remove

File: core/.gitignore
Comment (gameofpointers) [reply core/.gitignore:1]: no, mistake, good catch, will remove

File: network.env
Comment (gameofpointers) [reply network.env:1]: yea i agree, it was added by mistake, will be removed

File: params/bootnodes.go
Comment (gameofpointers) [reply params/bootnodes.go:1]: This is removed, please look at the subsequent commits, I just bought the whole params module and removed the files in the later commit that were not in use or not needed, so that we can track the delta

File: internal/build/archive.go
Comment (gameofpointers) [reply internal/build/archive.go:1]: this was needed for the build process i think, will see, if not needed I will remove

File: internal/build/download.go
Comment (gameofpointers) [reply internal/build/download.go:1]: above comment

File: internal/build/env.go
Comment (gameofpointers) [reply internal/build/env.go:1]: above comment

File: internal/build/gotool.go
Comment (gameofpointers) [reply internal/build/gotool.go:1]: above comment

File: internal/build/util.go
Comment (gameofpointers) [reply internal/build/util.go:1]: above comment

File: internal/build/pgp.go
Comment (gameofpointers) [reply internal/build/pgp.go:1]: above comment

File: build/ci.go (hunk 6 above)
Comment (gameofpointers) [reply build/ci.go:124]: This is the main thing we have to agree on, I added this from the original go-quai as a building process because, the way @alejoacosta74 had it would cause an import cycle, so this was a clean way to do the building process. I will see if i can not have this and still build, if not, You have to recommend an alternative, because i didnt invent this and just moved the logic from go-quai

File: cmd/utils/defaults.go (hunk 1 above)
Comment (robschleusner) [reply cmd/utils/defaults.go:76]: As we pull in more of go-quai, this file will eventually be empty - Only needed it since a lot of the imported flags needed default values and wanted to keep it in one spot so it is easy to remove

Comment (robschleusner) [reply cmd/utils/defaults.go:76]: Fixed in this PR: https://github.com/gameofpointers/go-quai/pull/36/files

File: cmd/utils/flags.go (hunk 3 above)
Comment (robschleusner) [reply cmd/utils/flags.go:625]: Have made the change in this PR 36 (https://github.com/gameofpointers/go-quai/pull/36/files). Here is the behavior now:

1. File path input: User inputs toml file and it scrapes the toml for the [coinbases] section and imports addresses from there
```
    Command: --coinbases /path/to/coinbases.toml
    TOML Content:
    [coinbases]
    "00" = "0xAddress00"
    "01" = "0xAddress01"
    ...
```

2. Address input: User inputs a string of shard to coinbase addresses
```
    Command: --coinbases "00=0xAddress00,01=0xAddress01"
```

3. Left blank: If user doesn't add the flag, then node will look in the directory specified by the `config-dir` flag for a `config.toml` file and parse the [coinbases] section

File: coinbases.json
Comment (robschleusner) [reply coinbases.json:1]: Removed in TOML implementation

File: Makefile (hunk 4 above)
Comment (wizeguyy) [reply Makefile:41]: Sorry, I didn't mean to say `go run`. I meant build and then run the compiled binary. Makefile is not for running. It is for building.

Config files and CLI args are how you modify the behavior of the binary once it runs.

Run scripts (if you want to add one) are a way to have complex logic determine when/how to run the binary. We do not need this today, but keep it in mind as an option

Comment (gameofpointers) [reply Makefile]: Removed it

File: cmd/utils/flags.go (hunk 3 above)
Comment (gameofpointers) [reply cmd/utils/flags.go]: I squashed this commit with the coinbase one

File: build/ci.go (hunk 6 above)
Comment (gameofpointers) [reply build/ci.go]: Changed the name to "build" to avoid confusion

File: internal/build/archive.go
Comment (gameofpointers) [reply internal/build/archive.go:1]: This is part of the linter set process, it downloads the linter and extracts and runs it, so we need few of these files, If we remove the linter we can get rid of some of these files

File: internal/build/pgp.go
Comment (gameofpointers) [reply internal/build/pgp.go:1]: This was not being used, so completely removed it

File: internal/build/util.go
Comment (gameofpointers) [reply internal/build/util.go:1]: This has some of the helper methods for building and extracting the git information to print it on start

File: cmd/utils/defaults.go (hunk 1 above)
Comment (gameofpointers) [reply cmd/utils/defaults.go]: Removed the defaults file

File: cmd/utils/cmd.go (hunk 7 above)
Comment (gameofpointers) [reply cmd/utils/cmd.go]: yeah, moved it to the quaiconfig folder

File: cmd/go-quai/root.go [hunk 8]
```diff
@@ -1,4 +1,4 @@
-package cmd
+package main
```
[the remark this answers is not in the document]
Comment (wizeguyy) [reply cmd/go-quai/root.go:1]: @gameofpointers thoughts? Keeping your commands in the `cmd` folder makes sense, IMO

**Review by wizeguyy — APPROVED** [review]
Alejo's comment about cmd organization needs consideration, but I'm fine with whatever you guys agree on with regard to that. This PR is good to go IMO

File: cmd/go-quai/root.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (gameofpointers) [reply cmd/go-quai/root.go:1]: @alejoacosta74 I havent removed it from the cmd folder, its inside

**Review by Djadih — APPROVED** [review]
Overall very nice. I saw a lot of things that were above and beyond the minimum of just making it work, however that also introduces some extra risk. One of those for example is hardcoding the ports and ontology to 1x1.

That said I didn't see anything "wrong" but I say that naturally cleanups should occur on the live codebase instead of halting this PR (which is blocking about 4 others) any further.

File: cmd/go-quai/root.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (Djadih) [reply cmd/go-quai/root.go:1]: @alejoacosta74 apparently cobra is tolerant to custom package structures.
> While you are welcome to provide your own organization, typically a Cobra-based application will follow the following organizational structure

There are some import cycles that happen with the original structure, and for sure its fixable, but it can also be done later and doesn't seem to be affecting cobra functionality.
